File: Data Structures & Algorithms/invert-a-binary-tree/submission-7.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {TreeNode}
     */
    invertTree(root) {
			if (!root) return null;
		const q = [root];
		while (q.length) {
			const top = q.shift();
			const right = top.right;
			top.right = top.left;
			top.left = right;
			if (top.right) q.push(top.right);
			if (top.left) q.push(top.left);
		}
		return root;

    }
}

File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-7.py
class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if not subRoot:
            return True
        if not root:
            return False
        if self.isSameTree(root, subRoot):
            return True
        return self.isSameTree(root.left, subRoot) or self.isSameTree(
            root.right, subRoot
        )

    def isSameTree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if not subRoot and not root:
            return True
        if root and subRoot and root.val == subRoot.val:
            return self.isSameTree(root.left, subRoot.left) and self.isSameTree(
                root.right, subRoot.right
            )
        return False
